src/main/java/com/ecore/roles/service/impl/MembershipsServiceImpl.java

```diff
@@ -8,48 +9,62 @@
 import com.ecore.roles.repository.MembershipRepository;
 import com.ecore.roles.repository.RoleRepository;
 import com.ecore.roles.service.MembershipsService;
+import com.ecore.roles.service.TeamsService;
 import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
 import lombok.extern.log4j.Log4j2;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import javax.validation.Valid;
 import java.util.List;
+import java.util.Objects;
 import java.util.UUID;
 
-import static java.util.Optional.ofNullable;
-
 @Log4j2
 @Service
+@RequiredArgsConstructor
 public class MembershipsServiceImpl implements MembershipsService {
 
     private final MembershipRepository membershipRepository;
     private final RoleRepository roleRepository;
-
-    @Autowired
-    public MembershipsServiceImpl(
-            MembershipRepository membershipRepository,
-            RoleRepository roleRepository) {
-        this.membershipRepository = membershipRepository;
-        this.roleRepository = roleRepository;
-    }
+    private final TeamsService teamsService;
 
     @Override
-    public Membership assignRoleToMembership(@NonNull Membership m) {
-
-        UUID roleId = ofNullable(m.getRole()).map(Role::getId)
-                .orElseThrow(() -> new InvalidArgumentException(Role.class));
+    public Membership assignRoleToMembership(@NonNull @Valid Membership membership) {
+        validateResourceExists(membership);
+        validateTeamMember(membership);
+        return membershipRepository.save(membership);
+    }
 
-        if (membershipRepository.findByUserIdAndTeamId(m.getUserId(), m.getTeamId())
-                .isPresent()) {
+    private void validateResourceExists(Membership membership) {
+        UUID roleId = membership.getRole().getId();
+        Role role = roleRepository.findById(roleId)
+                .orElseThrow(() -> new ResourceNotFoundException(Role.class, roleId));
+        if (membershipRepository.existsByUserIdAndTeamIdAndRole(membership.getUserId(),
```

Comment:
This new query has been created to validate business key conflict and throw ResourceExistsException if so.
There may be a small time window that two requests can insert the same business key then a DataIntegrityViolationException will be thrown because of the membership unique database index violation.